test/helpers.lua

```diff
@@ -17,27 +17,39 @@ local ok = function(res)
   return assert.is_true(res)
 end
 
+-- initial_path:  directory to recurse into
+-- re:            include pattern (string)
+-- exc_re:        exclude pattern(s) (string or table)
 local function glob(initial_path, re, exc_re)
+  exc_re = type(exc_re) == 'table' and exc_re or { exc_re }
   local paths_to_check = {initial_path}
   local ret = {}
   local checked_files = {}
+  local function is_excluded(path)
+    for _, pat in pairs(exc_re) do
+      if path:match(pat) then return true end
+    end
+    return false
+  end
+
   while #paths_to_check > 0 do
     local cur_path = paths_to_check[#paths_to_check]
     paths_to_check[#paths_to_check] = nil
-    for e in lfs.dir(cur_path) do
-      local full_path = cur_path .. '/' .. e
-      local checked_path = full_path:sub(#initial_path + 1)
-      if ((not exc_re or not checked_path:match(exc_re))
-          and e:sub(1, 1) ~= '.') then
-        local attrs = lfs.attributes(full_path)
-        if attrs then
-          local check_key = attrs.dev .. ':' .. tostring(attrs.ino)
-          if not checked_files[check_key] then
-            checked_files[check_key] = true
-            if attrs.mode == 'directory' then
-              paths_to_check[#paths_to_check + 1] = full_path
-            elseif not re or checked_path:match(re) then
-              ret[#ret + 1] = full_path
+    if lfs.attributes(cur_path) then  -- Path exists?
```

Comment:
Whether this path exists was already checked when it was first added (the only path not checked is the root of the search). Given that you use `lfs.attributes` and not `open()` with EAFP this is only going to make trying to access deleted path less likely, not fix anything. Not worth adding another level, use `pcall(lfs.dir, cur_dir)` if needed instead.